components/tidb_query_common/src/storage/scanner.rs

```diff
@@ -58,24 +97,26 @@ impl<T: Storage> RangesScanner<T> {
             },
             working_range_begin_key: Vec::with_capacity(KEY_BUFFER_CAPACITY),
             working_range_end_key: Vec::with_capacity(KEY_BUFFER_CAPACITY),
+            rescheduler: RescheduleChecker::new(),
         }
     }
 
     /// Fetches next row.
     // Note: This is not implemented over `Iterator` since it can fail.
     // TODO: Change to use reference to avoid allocation and copy.
-    pub fn next(&mut self) -> Result<Option<OwnedKvPair>, StorageError> {
-        self.next_opt(true)
+    pub async fn next(&mut self) -> Result<Option<OwnedKvPair>, StorageError> {
+        self.next_opt(true).await
     }
 
     /// Fetches next row.
     /// Note: `update_scanned_range` can control whether update the scanned
     /// range when `is_scanned_range_aware` is true.
-    pub fn next_opt(
+    pub async fn next_opt(
         &mut self,
         update_scanned_range: bool,
     ) -> Result<Option<OwnedKvPair>, StorageError> {
         loop {
+            let mut force_check = true;
```

Comment:
Nice finding. Didn't notice that iterator renew/re-seek is the cause of the long duration.